Add Writing of Single-Strip GeoTIFFs

DanielJDufour opened the pull request:
I added a `create` function, which takes in the values and metadata as two separate parameters. I hope this PR can serve to continue the conversation from https://github.com/constantinius/geotiff.js/issues/39

Let me know if there's anything that you'd like me to change.  Making this work within your current structure is a top priority for me.  I also changed ` if (pixelOffset < tileLength-1)` to ` if (pixelOffset < tileLength)` in line 454 of geotiffimage.js in order to read the last pixel of the tile.  I might be wrong though. Some more clarity on this logic would be appreciated :-)

I also wrote some tests for single-band rasters in geotiff.spec.js

Thanks again and look forward to your comments!

constantinius commented:
Hi @DanielJDufour 

Thank you for your contribution, I can see that you have put a lot of effort into this, which is very much appreciated. I think it is awesome, that it is already possible to write TIFF files!

I can see that the bulk of the new functionality is located in the `geotiffwriter.js`, which is a good idea in my opinion. I also think that having a really simple interface where some options can be tweaked is very important in order to create a usable library. The intent of geotiff.js is to hide as much of the gritty TIFF implementation details (tiled/stripped, pixel/band interleave, colorspaces, compression, data types, etc), and provide a unified access to the stuff that actually matter to the users: the actual raster data (+ metadata, potentially).

Ideally, the creation of GeoTIFFs works in the same way: the user only *has to* provide the raster data (raw typed arrays) + width and height and a TIFF is generated for him. If he wants, he should be able to specify what compression to be used, the tile size, the interleave method, etc. But geotiff.js should always handle the actual translation of data/metadata into the underlying format of tags and raw bytes. IMO the user should not be bothered with the necessity of actually having to specify the TIFF tags.

In this regard, I think your PR is definitely a good and promising starting point. I would propose that the next step would be to create a more concise interface function which allows a reasonable amount of options (with good default values) which are translated under the hood to the TIFF tags and so forth.

Otherwise, awesome!

> I also changed if (pixelOffset < tileLength-1) to if (pixelOffset < tileLength) in line 454 of geotiffimage.js in order to read the last pixel of the tile. I might be wrong though. Some more clarity on this logic would be appreciated :-)

This is not related to the writing of GeoTIFFs, right? For the sake of traceability, I think it is best to create a separate issue. So if you have some files (or generated one with your writing library) that do  not work because of the bug you mentioned please open a new issue, so that we don't get sidetracked and stay focused on the writing of geotiffs here.

DanielJDufour commented:
All good points! Thanks for the prompt feedback.

I'll work on integrating your feedback but it will probably take a few weeks. Sorry, just trying to set expectations.

Thanks, again!

DanielJDufour commented:
Hi, @constantinius , I just updated this branch with some changes.  Could you take a look, please?

I simplified the writing process, so the user only has to specify the pixel values, width and height.  It will default to a WGS 84 / 4326 raster covering the whole globe.

I also added support for RGB rasters.

I also created two ObservableHQ notebooks so you can see the code in action:
- https://beta.observablehq.com/@danieljdufour/create-an-rgb-geotiff 
- https://beta.observablehq.com/@danieljdufour/create-geotiff

As always, let me know if you'd like to change something.  I appreciate your insight on how to make this a good user experience :-)

Thanks, again and sorry for the delay!

sbilbault commented:
Hi, I was looking on a way to write geotiff in js from an array of number and metadata and found this PR which is exactly doing this ;-) Do you think this PR would be soon merged with this writting feature ? Thanks, Sébastien

DanielJDufour commented:
@sbilbault , I just fixed the merge conflicts, so you should be able to write GeoTIFFs if you want to use this branch.

@constantinius , could you take another look at this PR and let me know if there is anything else you would like to see?  Thank you!

DanielJDufour commented:
@sbilbault , did this PR work for you?  Were there any bugs or do you have any suggestions on how to improve it?  Thanks!

constantinius commented:
Hi @DanielJDufour, @sbilbault 

Sorry for responding not earlier, I did not get any notification, so I just stumbled upon these recent developments on accident. It's weird, as notifications are enabled.

Regarding the PR: it is definitely a nice feature to have and, as can be seen, there already is demand. I'll have to dig into the changes and see what you are actually doing.

From first glance, I'd say it looks solid, with some minor details to polish.

sbilbault commented:
Hey, the PR is working nicely, thanks a lot !

Le dim. 2 sept. 2018 à 17:13, Daniel J. Dufour <notifications@github.com> a
écrit :

> @sbilbault <https://github.com/sbilbault> , did this PR work for you?
> Were there any bugs or do you have any suggestions on how to improve it?
> Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geotiffjs/geotiff.js/pull/41#issuecomment-417937653>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AhV--Tnab3bpSWa1-yeaxj6EO-Re5aLsks5uW_WxgaJpZM4SLNLh>
> .
>

DanielJDufour commented:
Hi, @constantinius .  Just checking up on this.  Is there any more work that you would like me to do on this before the PR is merged?  Thanks!! :-)

constantinius commented:
@DanielJDufour 
I think it is fine to merge this. Tests and docs are fine, there are many helper functions in utils.js for my taste, but I think we can live with that.

Just a note (I think I hinted at it already): in the future, I might add additional options to encode images as TIFFs. For this, I'd like to use your code as a starting point. Is it okay for you when the files get modified in the future?

DanielJDufour commented:
@constantinius , that sounds good to me.  I'm definitely okay with the code being changed in the future! 
 Thank you :-)
